File: bindings/julia/src/SafeColor.jl
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeColor

Safe color operations with validated color space conversions.
"""
module SafeColor

export Rgb, Rgba, Hsl, Hsla
export parse_hex, parse_rgb, parse_hsl
export rgb_to_hex, rgba_to_hex, rgb_to_hsl, hsl_to_rgb
export format_rgb, format_rgba, format_hsl, format_hsla
export blend, lighten, darken, saturate, desaturate
export luminance, contrast_ratio, is_light, is_dark

"""
    Rgb

RGB color with values 0-255.
"""
struct Rgb
    r::UInt8
    g::UInt8
    b::UInt8
end

"""
    Rgba

RGBA color with alpha 0.0-1.0.
"""
struct Rgba
    r::UInt8
    g::UInt8
    b::UInt8
    a::Float64

    function Rgba(r::UInt8, g::UInt8, b::UInt8, a::Float64)
        new(r, g, b, clamp(a, 0.0, 1.0))
    end
end

Rgba(r, g, b, a) = Rgba(UInt8(r), UInt8(g), UInt8(b), Float64(a))
Rgba(rgb::Rgb, a::Float64=1.0) = Rgba(rgb.r, rgb.g, rgb.b, a)

"""
    Hsl

HSL color with h: 0-360, s: 0-100, l: 0-100.
"""
struct Hsl
    h::Float64  # 0-360
    s::Float64  # 0-100
    l::Float64  # 0-100

    function Hsl(h::Float64, s::Float64, l::Float64)
        new(mod(h, 360.0), clamp(s, 0.0, 100.0), clamp(l, 0.0, 100.0))
    end
end

Hsl(h, s, l) = Hsl(Float64(h), Float64(s), Float64(l))

"""
    Hsla

HSLA color with alpha 0.0-1.0.
"""
struct Hsla
    h::Float64
    s::Float64
    l::Float64
    a::Float64

    function Hsla(h::Float64, s::Float64, l::Float64, a::Float64)
        new(mod(h, 360.0), clamp(s, 0.0, 100.0), clamp(l, 0.0, 100.0), clamp(a, 0.0, 1.0))
    end
end

Hsla(h, s, l, a) = Hsla(Float64(h), Float64(s), Float64(l), Float64(a))
Hsla(hsl::Hsl, a::Float64=1.0) = Hsla(hsl.h, hsl.s, hsl.l, a)

"""
    parse_hex(s::AbstractString) -> Union{Rgb, Nothing}

Parse hex color string (#RGB, #RRGGBB).
"""
function parse_hex(s::AbstractString)::Union{Rgb, Nothing}
    str = strip(s)
    startswith(str, "#") && (str = str[2:end])

    if length(str) == 3
        # Short form #RGB
        r = tryparse(UInt8, string(str[1], str[1]); base=16)
        g = tryparse(UInt8, string(str[2], str[2]); base=16)
        b = tryparse(UInt8, string(str[3], str[3]); base=16)
    elseif length(str) == 6
        # Long form #RRGGBB
        r = tryparse(UInt8, str[1:2]; base=16)
        g = tryparse(UInt8, str[3:4]; base=16)
        b = tryparse(UInt8, str[5:6]; base=16)
    else
        return nothing
    end

    (r === nothing || g === nothing || b === nothing) && return nothing
    Rgb(r, g, b)
end

"""
    parse_rgb(s::AbstractString) -> Union{Rgb, Nothing}

Parse CSS rgb() format.
"""
function parse_rgb(s::AbstractString)::Union{Rgb, Nothing}
    str = lowercase(strip(s))

    m = match(r"^rgb\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)$", str)
    m === nothing && return nothing

    r = tryparse(Int, m.captures[1])
    g = tryparse(Int, m.captures[2])
    b = tryparse(Int, m.captures[3])

    (r === nothing || g === nothing || b === nothing) && return nothing
    (r < 0 || r > 255 || g < 0 || g > 255 || b < 0 || b > 255) && return nothing

    Rgb(UInt8(r), UInt8(g), UInt8(b))
end

"""
    parse_hsl(s::AbstractString) -> Union{Hsl, Nothing}

Parse CSS hsl() format.
"""
function parse_hsl(s::AbstractString)::Union{Hsl, Nothing}
    str = lowercase(strip(s))

    m = match(r"^hsl\s*\(\s*([\d.]+)\s*,\s*([\d.]+)%?\s*,\s*([\d.]+)%?\s*\)$", str)
    m === nothing && return nothing

    h = tryparse(Float64, m.captures[1])
    s_val = tryparse(Float64, m.captures[2])
    l = tryparse(Float64, m.captures[3])

    (h === nothing || s_val === nothing || l === nothing) && return nothing

    Hsl(h, s_val, l)
end

"""
    rgb_to_hex(c::Rgb) -> String

Convert RGB to hex string (#RRGGBB).
"""
function rgb_to_hex(c::Rgb)::String
    @sprintf("#%02x%02x%02x", c.r, c.g, c.b)
end

"""
    rgba_to_hex(c::Rgba) -> String

Convert RGBA to hex string (#RRGGBBAA).
"""
function rgba_to_hex(c::Rgba)::String
    alpha_byte = round(UInt8, c.a * 255)
    @sprintf("#%02x%02x%02x%02x", c.r, c.g, c.b, alpha_byte)
end

"""
    rgb_to_hsl(c::Rgb) -> Hsl

Convert RGB to HSL.
"""
function rgb_to_hsl(c::Rgb)::Hsl
    r = c.r / 255.0
    g = c.g / 255.0
    b = c.b / 255.0

    max_c = max(r, g, b)
    min_c = min(r, g, b)
    delta = max_c - min_c

    l = (max_c + min_c) / 2.0

    if delta == 0.0
        return Hsl(0.0, 0.0, l * 100.0)
    end

    s = if l < 0.5
        delta / (max_c + min_c)
    else
        delta / (2.0 - max_c - min_c)
    end

    h = if max_c == r
        60.0 * (((g - b) / delta) % 6)
    elseif max_c == g
        60.0 * (((b - r) / delta) + 2)
    else
        60.0 * (((r - g) / delta) + 4)
    end

    h < 0 && (h += 360.0)

    Hsl(h, s * 100.0, l * 100.0)
end

"""
    hsl_to_rgb(c::Hsl) -> Rgb

Convert HSL to RGB.
"""
function hsl_to_rgb(c::Hsl)::Rgb
    h = c.h / 360.0
    s = c.s / 100.0
    l = c.l / 100.0

    if s == 0.0
        v = round(UInt8, l * 255)
        return Rgb(v, v, v)
    end

    q = if l < 0.5
        l * (1.0 + s)
    else
        l + s - l * s
    end
    p = 2.0 * l - q

    function hue_to_rgb(t::Float64)::Float64
        t < 0 && (t += 1.0)
        t > 1 && (t -= 1.0)

        if t < 1/6
            p + (q - p) * 6.0 * t
        elseif t < 1/2
            q
        elseif t < 2/3
            p + (q - p) * (2/3 - t) * 6.0
        else
            p
        end
    end

    r = round(UInt8, hue_to_rgb(h + 1/3) * 255)
    g = round(UInt8, hue_to_rgb(h) * 255)
    b = round(UInt8, hue_to_rgb(h - 1/3) * 255)

    Rgb(r, g, b)
end

"""
    format_rgb(c::Rgb) -> String

Format as CSS rgb().
"""
format_rgb(c::Rgb)::String = "rgb($(c.r), $(c.g), $(c.b))"

"""
    format_rgba(c::Rgba) -> String

Format as CSS rgba().
"""
format_rgba(c::Rgba)::String = "rgba($(c.r), $(c.g), $(c.b), $(round(c.a, digits=2)))"

"""
    format_hsl(c::Hsl) -> String

Format as CSS hsl().
"""
format_hsl(c::Hsl)::String = "hsl($(round(c.h, digits=1)), $(round(c.s, digits=1))%, $(round(c.l, digits=1))%)"

"""
    format_hsla(c::Hsla) -> String

Format as CSS hsla().
"""
format_hsla(c::Hsla)::String = "hsla($(round(c.h, digits=1)), $(round(c.s, digits=1))%, $(round(c.l, digits=1))%, $(round(c.a, digits=2)))"

"""
    blend(c1::Rgb, c2::Rgb, factor::Float64) -> Rgb

Blend two colors. Factor 0.0 = c1, 1.0 = c2.
"""
function blend(c1::Rgb, c2::Rgb, factor::Float64)::Rgb
    f = clamp(factor, 0.0, 1.0)
    r = round(UInt8, c1.r * (1.0 - f) + c2.r * f)
    g = round(UInt8, c1.g * (1.0 - f) + c2.g * f)
    b = round(UInt8, c1.b * (1.0 - f) + c2.b * f)
    Rgb(r, g, b)
end

"""
    lighten(c::Rgb, amount::Float64) -> Rgb

Lighten a color by amount (0.0-1.0).
"""
function lighten(c::Rgb, amount::Float64)::Rgb
    hsl = rgb_to_hsl(c)
    new_l = min(100.0, hsl.l + amount * 100.0)
    hsl_to_rgb(Hsl(hsl.h, hsl.s, new_l))
end

"""
    darken(c::Rgb, amount::Float64) -> Rgb

Darken a color by amount (0.0-1.0).
"""
function darken(c::Rgb, amount::Float64)::Rgb
    hsl = rgb_to_hsl(c)
    new_l = max(0.0, hsl.l - amount * 100.0)
    hsl_to_rgb(Hsl(hsl.h, hsl.s, new_l))
end

"""
    saturate(c::Rgb, amount::Float64) -> Rgb

Increase saturation by amount (0.0-1.0).
"""
function saturate(c::Rgb, amount::Float64)::Rgb
    hsl = rgb_to_hsl(c)
    new_s = min(100.0, hsl.s + amount * 100.0)
    hsl_to_rgb(Hsl(hsl.h, new_s, hsl.l))
end

"""
    desaturate(c::Rgb, amount::Float64) -> Rgb

Decrease saturation by amount (0.0-1.0).
"""
function desaturate(c::Rgb, amount::Float64)::Rgb
    hsl = rgb_to_hsl(c)
    new_s = max(0.0, hsl.s - amount * 100.0)
    hsl_to_rgb(Hsl(hsl.h, new_s, hsl.l))
end

"""
    luminance(c::Rgb) -> Float64

Calculate relative luminance (0.0-1.0) per WCAG 2.1.
"""
function luminance(c::Rgb)::Float64
    function channel_luminance(v::UInt8)::Float64
        val = v / 255.0
        if val <= 0.03928
            val / 12.92
        else
            ((val + 0.055) / 1.055)^2.4
        end
    end

    0.2126 * channel_luminance(c.r) +
    0.7152 * channel_luminance(c.g) +
    0.0722 * channel_luminance(c.b)
end

"""
    contrast_ratio(c1::Rgb, c2::Rgb) -> Float64

Calculate WCAG contrast ratio between two colors.
"""
function contrast_ratio(c1::Rgb, c2::Rgb)::Float64
    l1 = luminance(c1)
    l2 = luminance(c2)

    lighter = max(l1, l2)
    darker = min(l1, l2)

    (lighter + 0.05) / (darker + 0.05)
end

"""
    is_light(c::Rgb) -> Bool

Check if color is light (luminance > 0.5).
"""
is_light(c::Rgb)::Bool = luminance(c) > 0.5

"""
    is_dark(c::Rgb) -> Bool

Check if color is dark (luminance <= 0.5).
"""
is_dark(c::Rgb)::Bool = luminance(c) <= 0.5

using Printf

# Display
Base.show(io::IO, c::Rgb) = print(io, rgb_to_hex(c))
Base.show(io::IO, c::Rgba) = print(io, format_rgba(c))
Base.show(io::IO, c::Hsl) = print(io, format_hsl(c))
Base.show(io::IO, c::Hsla) = print(io, format_hsla(c))

end # module

File: bindings/ocaml/lib/safe_color.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe color manipulation with clamped values. *)

(** RGB color with values 0-255. *)
type rgb = {
  r: int;
  g: int;
  b: int;
}

(** RGBA color with alpha 0.0-1.0. *)
type rgba = {
  r: int;
  g: int;
  b: int;
  a: float;
}

(** HSL color (hue 0-360, saturation/lightness 0.0-1.0). *)
type hsl = {
  h: float;
  s: float;
  l: float;
}

(** HSLA color with alpha. *)
type hsla = {
  h: float;
  s: float;
  l: float;
  a: float;
}

type error =
  | Invalid_format
  | Invalid_hex
  | Value_out_of_range

(** Clamp an integer to 0-255. *)
let clamp_byte v = max 0 (min 255 v)

(** Clamp a float to 0.0-1.0. *)
let clamp_unit v = Float.max 0.0 (Float.min 1.0 v)

(** Create RGB color with clamping. *)
let rgb r g b =
  { r = clamp_byte r; g = clamp_byte g; b = clamp_byte b }

(** Create RGBA color with clamping. *)
let rgba r g b a =
  { r = clamp_byte r; g = clamp_byte g; b = clamp_byte b; a = clamp_unit a }

(** Create HSL color with clamping. *)
let hsl h s l =
  let h = Float.rem h 360.0 in
  let h = if h < 0.0 then h +. 360.0 else h in
  { h; s = clamp_unit s; l = clamp_unit l }

(** Create HSLA color with clamping. *)
let hsla h s l a =
  let h = Float.rem h 360.0 in
  let h = if h < 0.0 then h +. 360.0 else h in
  { h; s = clamp_unit s; l = clamp_unit l; a = clamp_unit a }

(** Parse hex color string (#RGB, #RRGGBB, #RGBA, #RRGGBBAA). *)
let parse_hex str =
  let str = String.trim str in
  let str = if String.length str > 0 && str.[0] = '#' then
    String.sub str 1 (String.length str - 1)
  else str in
  let parse_byte hex =
    try int_of_string ("0x" ^ hex)
    with _ -> -1
  in
  let len = String.length str in
  if len = 3 || len = 4 then
    (* Short form: #RGB or #RGBA *)
    let r = parse_byte (String.make 2 str.[0]) in
    let g = parse_byte (String.make 2 str.[1]) in
    let b = parse_byte (String.make 2 str.[2]) in
    if r < 0 || g < 0 || b < 0 then Error Invalid_hex
    else if len = 3 then
      Ok (rgba r g b 1.0)
    else
      let a = parse_byte (String.make 2 str.[3]) in
      if a < 0 then Error Invalid_hex
      else Ok (rgba r g b (float_of_int a /. 255.0))
  else if len = 6 || len = 8 then
    (* Long form: #RRGGBB or #RRGGBBAA *)
    let r = parse_byte (String.sub str 0 2) in
    let g = parse_byte (String.sub str 2 2) in
    let b = parse_byte (String.sub str 4 2) in
    if r < 0 || g < 0 || b < 0 then Error Invalid_hex
    else if len = 6 then
      Ok (rgba r g b 1.0)
    else
      let a = parse_byte (String.sub str 6 2) in
      if a < 0 then Error Invalid_hex
      else Ok (rgba r g b (float_of_int a /. 255.0))
  else
    Error Invalid_format

(** Format RGB as hex string. *)
let to_hex color =
  Printf.sprintf "#%02x%02x%02x" color.r color.g color.b

(** Format RGBA as hex string with alpha. *)
let to_hex_rgba color =
  let alpha_byte = int_of_float (color.a *. 255.0) in
  Printf.sprintf "#%02x%02x%02x%02x" color.r color.g color.b alpha_byte

(** Format RGB as CSS rgb() function. *)
let to_css_rgb color =
  Printf.sprintf "rgb(%d, %d, %d)" color.r color.g color.b

(** Format RGBA as CSS rgba() function. *)
let to_css_rgba color =
  Printf.sprintf "rgba(%d, %d, %d, %.3f)" color.r color.g color.b color.a

(** Format HSL as CSS hsl() function. *)
let to_css_hsl color =
  Printf.sprintf "hsl(%.1f, %.1f%%, %.1f%%)"
    color.h (color.s *. 100.0) (color.l *. 100.0)

(** Convert RGB to HSL. *)
let rgb_to_hsl color =
  let r = float_of_int color.r /. 255.0 in
  let g = float_of_int color.g /. 255.0 in
  let b = float_of_int color.b /. 255.0 in
  let max_c = Float.max r (Float.max g b) in
  let min_c = Float.min r (Float.min g b) in
  let l = (max_c +. min_c) /. 2.0 in
  if max_c = min_c then
    { h = 0.0; s = 0.0; l }
  else
    let d = max_c -. min_c in
    let s = if l > 0.5 then d /. (2.0 -. max_c -. min_c) else d /. (max_c +. min_c) in
    let h =
      if max_c = r then
        (g -. b) /. d +. (if g < b then 6.0 else 0.0)
      else if max_c = g then
        (b -. r) /. d +. 2.0
      else
        (r -. g) /. d +. 4.0
    in
    { h = h *. 60.0; s; l }

(** Convert HSL to RGB. *)
let hsl_to_rgb color =
  let h = color.h /. 360.0 in
  let s = color.s in
  let l = color.l in
  if s = 0.0 then
    let v = int_of_float (l *. 255.0) in
    { r = v; g = v; b = v }
  else
    let hue_to_rgb p q t =
      let t = if t < 0.0 then t +. 1.0 else if t > 1.0 then t -. 1.0 else t in
      if t < 1.0 /. 6.0 then p +. (q -. p) *. 6.0 *. t
      else if t < 1.0 /. 2.0 then q
      else if t < 2.0 /. 3.0 then p +. (q -. p) *. (2.0 /. 3.0 -. t) *. 6.0
      else p
    in
    let q = if l < 0.5 then l *. (1.0 +. s) else l +. s -. l *. s in
    let p = 2.0 *. l -. q in
    let r = hue_to_rgb p q (h +. 1.0 /. 3.0) in
    let g = hue_to_rgb p q h in
    let b = hue_to_rgb p q (h -. 1.0 /. 3.0) in
    { r = int_of_float (r *. 255.0);
      g = int_of_float (g *. 255.0);
      b = int_of_float (b *. 255.0) }

(** Lighten a color by a percentage (0.0-1.0). *)
let lighten amount color =
  let h = rgb_to_hsl color in
  let new_l = clamp_unit (h.l +. amount) in
  hsl_to_rgb { h with l = new_l }

(** Darken a color by a percentage (0.0-1.0). *)
let darken amount color =
  let h = rgb_to_hsl color in
  let new_l = clamp_unit (h.l -. amount) in
  hsl_to_rgb { h with l = new_l }

(** Saturate a color by a percentage. *)
let saturate amount color =
  let h = rgb_to_hsl color in
  let new_s = clamp_unit (h.s +. amount) in
  hsl_to_rgb { h with s = new_s }

(** Desaturate a color by a percentage. *)
let desaturate amount color =
  let h = rgb_to_hsl color in
  let new_s = clamp_unit (h.s -. amount) in
  hsl_to_rgb { h with s = new_s }

(** Rotate hue by degrees. *)
let rotate_hue degrees color =
  let h = rgb_to_hsl color in
  let new_h = Float.rem (h.h +. degrees) 360.0 in
  let new_h = if new_h < 0.0 then new_h +. 360.0 else new_h in
  hsl_to_rgb { h with h = new_h }

(** Get complementary color (180 degree rotation). *)
let complement color =
  rotate_hue 180.0 color

(** Invert a color. *)
let invert color =
  { r = 255 - color.r; g = 255 - color.g; b = 255 - color.b }

(** Grayscale conversion using luminosity method. *)
let grayscale color =
  let gray = int_of_float (
    float_of_int color.r *. 0.299 +.
    float_of_int color.g *. 0.587 +.
    float_of_int color.b *. 0.114
  ) in
  { r = gray; g = gray; b = gray }

(** Mix two colors with a weight (0.0 = all first, 1.0 = all second). *)
let mix weight c1 c2 =
  let w = clamp_unit weight in
  let mix_channel a b = int_of_float (
    float_of_int a *. (1.0 -. w) +. float_of_int b *. w
  ) in
  { r = mix_channel c1.r c2.r;
    g = mix_channel c1.g c2.g;
    b = mix_channel c1.b c2.b }

(** Calculate contrast ratio between two colors (WCAG). *)
let contrast_ratio c1 c2 =
  let relative_luminance c =
    let channel v =
      let v = float_of_int v /. 255.0 in
      if v <= 0.03928 then v /. 12.92
      else Float.pow ((v +. 0.055) /. 1.055) 2.4
    in
    0.2126 *. channel c.r +. 0.7152 *. channel c.g +. 0.0722 *. channel c.b
  in
  let l1 = relative_luminance c1 in
  let l2 = relative_luminance c2 in
  let lighter = Float.max l1 l2 in
  let darker = Float.min l1 l2 in
  (lighter +. 0.05) /. (darker +. 0.05)

(** Check if colors have sufficient contrast for text (WCAG AA: 4.5:1). *)
let is_readable ?(level=4.5) c1 c2 =
  contrast_ratio c1 c2 >= level

(** Get a readable text color (black or white) for a background. *)
let text_color_for_background bg =
  let luminance =
    float_of_int bg.r *. 0.299 +.
    float_of_int bg.g *. 0.587 +.
    float_of_int bg.b *. 0.114
  in
  if luminance > 128.0 then { r = 0; g = 0; b = 0 }
  else { r = 255; g = 255; b = 255 }

(** Common colors. *)
let black = { r = 0; g = 0; b = 0 }
let white = { r = 255; g = 255; b = 255 }
let red = { r = 255; g = 0; b = 0 }
let green = { r = 0; g = 128; b = 0 }
let blue = { r = 0; g = 0; b = 255 }
let yellow = { r = 255; g = 255; b = 0 }
let cyan = { r = 0; g = 255; b = 255 }
let magenta = { r = 255; g = 0; b = 255 }
let orange = { r = 255; g = 165; b = 0 }
let purple = { r = 128; g = 0; b = 128 }
let pink = { r = 255; g = 192; b = 203 }
let gray = { r = 128; g = 128; b = 128 }
let transparent = { r = 0; g = 0; b = 0; a = 0.0 }
